rct229/rulesets/ashrae9012019/section16/section16rule2.py

```diff
@@ -0,0 +1,127 @@
+from rct229.rule_engine.rule_base import RuleDefinitionBase
+from rct229.rule_engine.rule_list_indexed_base import RuleDefinitionListIndexedBase
+from rct229.rule_engine.ruleset_model_factory import produce_ruleset_model_instance
+from rct229.rule_engine.rulesets import LeapYear
+from rct229.rulesets.ashrae9012019 import BASELINE_0
+from rct229.rulesets.ashrae9012019.ruleset_functions.compare_schedules import (
+    compare_schedules,
+)
+from rct229.utils.jsonpath_utils import find_all
+from rct229.utils.utility_functions import find_exactly_one_schedule
+
+
+class Section16Rule2(RuleDefinitionListIndexedBase):
+    """Rule 2 of ASHRAE 90.1-2019 Appendix G Section 16 (Elevators)"""
+
+    def __init__(self):
+        super(Section16Rule2, self).__init__(
+            rmds_used=produce_ruleset_model_instance(
+                USER=False,
+                BASELINE_0=True,
+                PROPOSED=True,
+            ),
+            each_rule=Section16Rule2.RuleSetModelDescriptionRule(),
+            index_rmd=BASELINE_0,
+            id="16-2",
+            description="The baseline elevator motor use shall be modeled with the same schedule as the proposed design. Rule Assertion: B-RMD = P-RMD",
+            ruleset_section_title="Elevators",
+            standard_section="Section G3.1",
+            is_primary_rule=True,
+            list_path="ruleset_model_descriptions[0]",
+            required_fields={"$": ["calendar"], "$.calendar": ["is_leap_year"]},
+            data_items={"is_leap_year_b": (BASELINE_0, "calendar/is_leap_year")},
+        )
+
+    class RuleSetModelDescriptionRule(RuleDefinitionListIndexedBase):
+        def __init__(self):
+            super(Section16Rule2.RuleSetModelDescriptionRule, self).__init__(
+                rmds_used=produce_ruleset_model_instance(
+                    USER=False, BASELINE_0=True, PROPOSED=True
+                ),
+                each_rule=Section16Rule2.RuleSetModelDescriptionRule.ElevatorRule(),
+                index_rmd=BASELINE_0,
+                list_path="buildings[*].elevators[*]",
+            )
```

Comment:
@JacksonJ-KC , could you check what outcome should it be if the baseline has no elevator but proposed has?
The rule passes if no elevator in baseline model but proposed has.

Reply:
This should not be a Pass. This would result in a fail for rule 1-6 already. My only questions is: do we want to provide another Fail outcome for this rule, or have it be not applicable?

Reply:
> This should not be a Pass. This would result in a fail for rule 1-6 already. My only questions is: do we want to provide another Fail outcome for this rule, or have it be not applicable?

It seems like an outcome at RPD level instead of elevator level. However, if 1-6 captures the mismatch, I am considering a not applicable outcome here as well. Since the issue is highlighting the missing elevators in baseline?

Reply:
For record purpose.

Email: May 31, 2024, Title: RCT, Question about section 16-2 rule

Weili:
"""
The key issue is not elevator key doesn’t exist but if the elevator is an empty list in the baseline model.
In that case, currently, the RCT will loop in an empty list and return nothing.

We will need to flag the issue if the baseline has empty list. There are two questions here:
if baseline has no elevator key or an empty elevator list but proposed has elevator(s), do we flag the rule as undermined or failed?
if proposed has no elevator but baseline has elevator(s), do we flag it not applicable, or failed, or undermined?
"""

Jackson:
"""Personally, I am leaning toward having the rule be not applicable for both scenarios and letting the Baseline == Proposed rule (1-6) handle the failed outcome that would result.
"""

Weili:
"""
I agree with that decision. I also left the same comment following your quote.
In that case, we should update all 16 RDSs to include this logic in the applicable evaluation. Jackson, could you update them all in one PR?

Yun Joon, Jiarong, please update section 16 rule applicable checks following the logic:

If proposed has no elevators OR (proposed has elevators AND (baseline has no elevators OR empty elevator list)), return NOT_APPLICABLE
"""

Reply:
@Jiarongx-Xie FYI

Reply:
Logic derivation:

P: elevator in proposed, B: elevator in baseline

Original  logic from Weili: (P is False) or (P is True and (B is None or B is False))

we should negate the original logic b/c we have the is_applicable (positive).

Applying De Morgan's laws
not (P is False) and not (P is True and (B is None or B is False))
= (P is True) and (P is False or (B is not None and B is True))
= P is True and (B is not None and B is True)

Reply:
@jugonzal07 Should we add a not_applicable case in the rpd?

Reply:
I do not think the De Morgan's law applied correctly here.
In the second step: `= (P is True) and (P is False or (B is not None and B is True))` it is impossible

Technically, the original logic should be:
`(P is False or (B is None or B is False))`

The logic after the first `or` has a pre-condition that P is True.

You can keep the original logic as it is for readability.

Reply:
Good point! For readability, I put ```not``` at the beginning of the logic.
When I tested the following 3 scenarios, the logic detects the not applicable cases correctly.
1) When the ```elevator_p``` doesn't exist
2) When the ```elevator_b``` doesn't exist
3) When the ```elevator_b``` is an empty list

Please let me know if any further changes need to be made.

Reply:
make sense to me.